### PR TITLE
Add tpm e2e tests

### PR DESCRIPTION
**Pull request by jinghanm**
## Description
Add Tpm E2E tests

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [x] All unit tests are passing.
- [x] I have merged the latest `main` branch prior to this PR submission.
- [x] I submitted this PR against the `main` branch.

### CODE REVIEW

**Comment by MariusNi** [comment]
@hachotze is added to the review. #Closed

File: src/tests/e2etest/TpmTests.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                TpmVersionPattern = new Regex(@"((\d+)\.)?((\d+)\.)?((\d+)\.(\*|\d+))"),
+                TpmManufacturerPattern = new Regex(@"[A-Za-z]([0-9a-zA-Z\-]+)(\s){0,1}([0-9a-zA-Z\-]+)?"),
+                TpmStatusPattern = new Regex(@"[0-2]")
+            };
+            var deserializedReportedObject = JsonSerializer.Deserialize<Tpm>(GetTwin().Properties.Reported[ComponentName].ToString());
+
+            if ((GetTwin().ConnectionState == DeviceConnectionState.Disconnected) || (GetTwin().Status == DeviceStatus.Disabled))
+            {
+                Assert.Fail("Module is disconnected or is disabled");
```
Comment (hachotze) [inline src/tests/e2etest/TpmTests.cs:48]: Consider changing "Module is disconnected..." to "Module twin is disconnected..." or "Twin is disconnected"

File: src/tests/e2etest/TpmTests.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+           Assert.True(IsRegexMatch(expectedTpmPattern.TpmStatusPattern, deserializedReportedObject.TpmStatus));
+           Assert.True(String.IsNullOrEmpty(deserializedReportedObject.TpmVersion) || IsRegexMatch(expectedTpmPattern.TpmVersionPattern, deserializedReportedObject.TpmVersion));
+           Assert.True(String.IsNullOrEmpty(deserializedReportedObject.TpmVersion) || IsRegexMatch(expectedTpmPattern.TpmManufacturerPattern, deserializedReportedObject.TpmManufacturer));
```
Comment (hachotze) [inline src/tests/e2etest/TpmTests.cs:53]: Should this be deserializedReportedObject.TpmManufacturer?

**Review by hachotze — COMMENTED** [review]
Looks great overall! I left a couple of suggestions and a question.

File: src/tests/e2etest/TpmTests.cs (hunk 1 above)
Comment (MariusNi) [reply src/tests/e2etest/TpmTests.cs:48]: No, this actually about what is in the Twin about the Module that owns the Twin.

File: src/tests/e2etest/TpmTests.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            public Regex TpmManufacturerPattern { get; set; }
+            public Regex TpmStatusPattern { get; set; }
+        }
+
+        [Test]
+        public void TpmTest()
+        {
+            var expectedTpmPattern = new ExpectedTpmPattern
+            {
+                TpmVersionPattern = new Regex(@"((\d+)\.)?((\d+)\.)?((\d+)\.(\*|\d+))"),
```
[the remark this answers is not in the document]
Comment (MariusNi) [reply src/tests/e2etest/TpmTests.cs:40]: ... also 2.x and y.z in the future?

File: src/tests/e2etest/TpmTests.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+using System.Text.Json;
+using System;
+using System.Threading.Tasks;
+using System.Text.RegularExpressions;
+
+namespace e2etesting
+{
+    public class TpmTests : E2eTest
+    {
+        string ComponentName = "Tpm";
```
Comment (AhmedBM) [inline src/tests/e2etest/TpmTests.cs:15]: Should be `readonly`

**Review by AhmedBM — APPROVED** [review]
Looks good! Also tested on E2E Cloud Tests pipeline 👌

File: src/tests/e2etest/TpmTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jinghanm) [reply src/tests/e2etest/TpmTests.cs:40]: Thanks for the information. I have fixed the regex.

File: src/tests/e2etest/TpmTests.cs (hunk 2 above)
Comment (jinghanm) [reply src/tests/e2etest/TpmTests.cs:53]: Thanks for pointing this out. I have fixed it.

File: src/tests/e2etest/TpmTests.cs (hunk 1 above)
Comment (jinghanm) [reply src/tests/e2etest/TpmTests.cs:48]: The twin is the image of a module (or device), I think it is module or device that is connected or disconnected?

**Review by hachotze — DISMISSED** [review]
Looks great, signing off.

**Review by MariusNi — APPROVED** [review]
:shipit:

File: src/tests/e2etest/TpmTests.cs (hunk 4 above)
Comment (jinghanm) [reply src/tests/e2etest/TpmTests.cs:15]: Thanks for pointing this out. Fixed!

**Review by AhmedBM — APPROVED** [review]

**Pull request merged**